### ISSUE TITLE
nullptr check and abort build

### ISSUE THREAD

**Issue by jonnygrant** [isocpp/CppCoreGuidelines#2071]
not_null is useful, it rejects nullptr at runtime and calls std::terminate()

I wanted something at build-time like a static_assert to find null pointers, so I customized a not_null container to reject any nullptr at build time in GCC as an example:
https://godbolt.org/z/x7xv7bWea

```
In constructor 'constexpr gsl::not_null<T>::not_null(T) [with <template-parameter-2-1> = void; T = int*]',
    inlined from 'void foo(int*)' at <source>:483:30:
<source>:240:43: error: call to 'gsl::compile_error_null_check' declared with attribute error: not_null found nullptr
  240 |         if(!ptr_) compile_error_null_check();
      |                   ~~~~~~~~~~~~~~~~~~~~~~~~^~
Compiler returned: 1
```

This came out of experiments discussed on [GSL ticket](https://github.com/microsoft/GSL/issues/1102) 

Wanting to hear other opinions and see if it's something (or similar) that could be added to CppCoreGuidelines. No problem if it isn't something GSL needs.  Could a _not_null_build<T*>_ be added?

My modification supports rule ES.65: Don’t dereference an invalid pointer
It enforces code to check for nullptr, and therefore not pass it.
Gives a useful Enforcement build error utilizing compiler optimizer.

My idea gives the programmer chance to evaluate where the nullptr came from, and what logic caused it. [rather than just put a condition like **_if(!p) return;_**]

Note, it would be possible to extend my approach to have a similar never_negative<T> that aborted builds when a never_negative<int> was passed -1. Of course can be anything, even a build time **_min_max_range_check<T, A, B>_**
or a _**never_empty_str<T*>**_. I've not tried with complex containers, but even _**never_empty_std_vector<T>**_ is probably possible.

I like bulletproof C++ code, that doesn't crash, and detects issues at compile time so this would really help!

**Comment by hsutter** [comment]
Editors call: We are interested in defending against null dereference, but don't want to be inventive right now beyond what `not_null` currently does.

But we'd like to ask @jwakely : Jonathan, do you have insights about this, and the related examples at https://github.com/microsoft/GSL/issues/1102 ?

**Comment by hsutter** [comment]
Thanks again for this suggestion.

We've discussed this and don't think this is a direction to pursue at this time. The suggested null detection appears to be reliant on optimization and inter-procedural analysis, which are nondeterministic and don't scale, respectively.

There can definitely be value in a "works sometimes, for simple cases only" solution, and we encourage you to do that in your own code if it works for you. We just don't think it's a high enough bar to put in the Guidelines or GSL to be reliable enough to mandate for everyone to use.

**Comment by jonnygrant** [comment]
I wanted to express my gratitude for taking the time to consider the container suggestion I proposed. The implementation relies heavily on control flow analysis, specifically dead code removal in the Optimizer. Consequently, it consistently requires at least -O1, mirroring the behavior observed with other advanced compiler-detected errors.

In case it proves beneficial, the implementation is readily available for those who seek a compile-time sanity check, particularly for security purposes. I've successfully deployed it in work projects to enforce API interface sanity checks for nullptr at build-time.

Considering the ongoing developments in programming languages, I can't help but speculate that C++ Concepts might find a similar approach valuable. Specifically, utilizing compiler analysis of control flow to satisfy a Precondition prerequisite, such as ensuring a pointer can't be the nullptr constant, it would require the caller to check it wasn't passing nullptr, which can be issued as a compiler warning if not.

Once again, thank you for the time spent considering.
